### ISSUE TITLE
Download build artifacts in Release deployment fails: ArtifactEngine.js unhandled rejection due to "read ECONNRESET"

### ISSUE THREAD

**Issue by Shakevg** [microsoft/azure-pipelines-tasks#10702]
## Information
**Bug**  

**Task Name**: DownloadBuildArtifactsV0

## Environment
    - Azure Release Pipelines (Release definitionId=209)
Example: https://dev.azure.com/pdd-ihsmarkit/PLC/_releaseProgress?_a=release-environment-logs&releaseId=20380&environmentId=77414

- Agent
    - Release Private
    - Agent running in Interactive mode
    - Queue name: Test Automation
    - Agent Version and Platform
vsts-agent-win-x64-2.153.1
Windows Server 2016 Standard 

## Description

This error occurs when running the Azure Dev Ops release built-in Download Artifacts from a build pipeline, artifactEngine.js throws undhandled rejection due to "read ECONNRESET".
**Then Release Agent became not workable and need to reboot it to restore (it is in active state, but all jobs failed on it).**

There is no information in the regular logs if there was a retry or not to handle this problem.

We have about 30 of release Agents and most of them succeed; but often the engine threw the exception.

Looks like similar to https://github.com/Microsoft/azure-pipelines-extensions/issues/597

### Task logs
```
==============================================================================
Task         : Download build artifacts
Description  : Download files that were saved as artifacts of a completed build
Version      : 0.153.1
Author       : Microsoft Corporation
==============================================================================
_work\_tasks\DownloadBuildArtifacts_a433f589-fce1-4460-9ee6-44a624aeb1fb\0.153.1\node_modules\artifact-engine\Engine\artifactEngine.js:157
##[error]artifact-engine: unhandled rejection Error: read ECONNRESET
     throw err;
     ^
 
Error: read ECONNRESET
     at exports._errnoException (util.js:1018:11)
     at TLSWrap.onread (net.js:568:26)
 ##[error]Unhandled: read ECONNRESET
 ##[error]Unhandled: read ECONNRESET
 ##[error]artifact-engine: unhandled exception Error: read ECONNRESET
 ##[error]Exit code 7 returned from process: file name 'externals\node\bin\node.exe', arguments '"_work\_tasks\DownloadBuildArtifacts_a433f589-fce1-4460-9ee6-44a624aeb1fb\0.153.1\main.js"'.
```

**Comment by huntleyh-MSFT** [comment]
This issue is happening for my customers as well. We've updated to the latest build agent v2.153.3 to ensure that's not the issue.
Our initial report for this resulted in the commit change: https://github.com/microsoft/azure-pipelines-tasks/commit/96494cc125378bc183180cd00fe472e655255410#diff-3838eb12f4eeaf3635c031b2de8ad1a6
However there's a new commit mentioning this has been reverted: [1c5e3e00c9e673289a21dce5a791ad4bc6fe9071#diff-3838eb12f4eeaf3635c031b2de8ad1a6](url)
Any ETA on resolving this?

**Comment by gykuma** [comment]
Hi @Shakevg , we recently releases a hotfix for this issue. Let us know if it fixed the issue for you.

**Comment by Shakevg** [comment]
Hi, @gykuma Download artifact task still failed often but error is different. Should I submit new issue or it is possible to process it is here?

Azure release: https://dev.azure.com/pdd-ihsmarkit/PLC/_releaseProgress?releaseId=23172&_a=release-environment-logs&environmentId=84609
Task logs (full log in attachment
[Download artifact task logs.txt](https://github.com/microsoft/azure-pipelines-tasks/files/3338237/Download.artifact.task.logs.txt)
):
```
2019-06-27T23:25:02.3879290Z ##[debug]Evaluating condition for step: 'Download artifact - EWB.RunUITests_RD - drop'
2019-06-27T23:25:02.3980408Z ##[debug]Evaluating: succeeded()
2019-06-27T23:25:02.3995377Z ##[debug]Evaluating succeeded:
2019-06-27T23:25:02.4033586Z ##[debug]=> True
2019-06-27T23:25:02.4058799Z ##[debug]Result: True
2019-06-27T23:25:02.4134463Z ##[section]Starting: Download artifact - EWB.RunUITests_RD - drop
2019-06-27T23:25:02.4971191Z ==============================================================================
2019-06-27T23:25:02.4971349Z Task         : Download build artifacts
2019-06-27T23:25:02.4971432Z Description  : Download files that were saved as artifacts of a completed build
2019-06-27T23:25:02.4971523Z Version      : 0.154.2
2019-06-27T23:25:02.4971590Z Author       : Microsoft Corporation
2019-06-27T23:25:02.4971673Z Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/download-build-artifacts
2019-06-27T23:25:02.4972024Z ==============================================================================
2019-06-27T23:25:02.8195270Z ##[debug]agent.TempDirectory=C:\vsts-agent-win-x64-2.144.0\_work\_
[Download artifact task logs.txt](https://github.com/microsoft/azure-pipelines-tasks/files/3338234/Download.artifact.task.logs.txt)
temp
2019-06-27T23:25:02.8195882Z ##[debug]loading inputs and endpoints
2019-06-27T23:25:02.8197252Z ##[debug]loading ENDPOINT_AUTH_PARAMETER_SYSTEMVSSCONNECTION_ACCESSTOKEN
2019-06-27T23:25:02.8197560Z ##[debug]loading ENDPOINT_AUTH_SCHEME_SYSTEMVSSCONNECTION
2019-06-27T23:25:02.8197801Z ##[debug]loading ENDPOINT_AUTH_SYSTEMVSSCONNECTION
2019-06-27T23:25:02.8198014Z ##[debug]loading INPUT_ALLOWPARTIALLYSUCCEEDEDBUILDS
2019-06-27T23:25:02.8198247Z ##[debug]loading INPUT_ARTIFACTNAME
2019-06-27T23:25:02.8198464Z ##[debug]loading INPUT_BRANCHNAME
2019-06-27T23:25:02.8198669Z ##[debug]loading INPUT_BUILDID
2019-06-27T23:25:02.8198882Z ##[debug]loading INPUT_BUILDTYPE
2019-06-27T23:25:02.8199103Z ##[debug]loading INPUT_BUILDVERSIONTODOWNLOAD
2019-06-27T23:25:02.8199340Z ##[debug]loading INPUT_DEFINITION
2019-06-27T23:25:02.8199581Z ##[debug]loading INPUT_DOWNLOADPATH
2019-06-27T23:25:02.8199822Z ##[debug]loading INPUT_DOWNLOADTYPE
2019-06-27T23:25:02.8200019Z ##[debug]loading INPUT_ITEMPATTERN
2019-06-27T23:25:02.8200232Z ##[debug]loading INPUT_PARALLELIZATIONLIMIT
2019-06-27T23:25:02.8200449Z ##[debug]loading INPUT_PROJECT
2019-06-27T23:25:02.8200647Z ##[debug]loading INPUT_SPECIFICBUILDWITHTRIGGERING
2019-06-27T23:25:02.8216039Z ##[debug]loading SECRET_PASS
2019-06-27T23:25:02.8219007Z ##[debug]loading SECRET_PASSIMB
2019-06-27T23:25:02.8221767Z ##[debug]loading SECRET_RMPASS
2019-06-27T23:25:02.8224731Z ##[debug]loading SECRET_RMUSER
2019-06-27T23:25:02.8227684Z ##[debug]loading SECRET_STAGEPASS
2019-06-27T23:25:02.8230566Z ##[debug]loading SECRET_TFSUSERACCESSTOKENVG
2019-06-27T23:25:02.8246838Z ##[debug]loaded 22
2019-06-27T23:25:02.8270523Z ##[debug]Agent.ProxyUrl=undefined
2019-06-27T23:25:02.8272126Z ##[debug]Agent.CAInfo=undefined
2019-06-27T23:25:02.8272446Z ##[debug]Agent.ClientCert=undefined
2019-06-27T23:25:02.8272837Z ##[debug]Agent.SkipCertValidation=undefined
....
2019-06-27T23:27:00.2471656Z ##[debug]task result: Failed
2019-06-27T23:27:00.2536424Z ##[error]Error: read ECONNRESET
2019-06-27T23:27:00.2552558Z ##[debug]Processed: ##vso[task.issue type=error;]Error: read ECONNRESET
2019-06-27T23:27:00.2592244Z ##[debug]Processed: ##vso[task.complete result=Failed;]Error: read ECONNRESET
2019-06-27T23:27:00.2592717Z ##[debug]Error getting drop/IHS.EWB.Test/Properties:Error: socket hang up
2019-06-27T23:27:00.2593021Z ##[debug]Error getting drop/IHS.CodeCoverageCollector/bin:Error: socket hang up
2019-06-27T23:27:00.5013577Z ##[debug]Aborting respawning worker, as download failed for some file(s).
2019-06-27T23:27:00.6730759Z ##[debug]Aborting respawning worker, as download failed for some file(s).
2019-06-27T23:27:00.7510535Z ##[debug]Aborting respawning worker, as download failed for some file(s).
2019-06-27T23:27:00.7511056Z ##[debug]Aborting respawning worker, as download failed for some file(s).
2019-06-27T23:27:00.9543825Z ##[debug]Aborting respawning worker, as download failed for some file(s).
2019-06-27T23:27:04.1419070Z ##[debug]Total Items: 29, Processed: 26, Processing: 0, Queued: 0, Skipped: 0, Failed: 3, Time elapsed: 120.22secs


```

**Comment by gykuma** [comment]
Hi @Shakevg we have added retries for handling these ECONNRESET related issues in the latest task version (0.155.1). We are closing this issue for now. Please feel free to reopen it if the problem still exists for you.

**Comment by Shakevg** [comment]
@gykuma When this version of Release Agent will be available on Azure (latest version is 2.154.3)?

**Comment by gykuma** [comment]
@Shakevg The version I mentioned is for the **Download Build Artifacts** task not the agent. You can check the latest task version available on your account by using this api `https://dev.azure.com/{{your-account-name}}/_apis/distributedtask/tasks/a433f589-fce1-4460-9ee6-44a624aeb1fb`
